lib/ansible/executor/task_executor.py

```diff
@@ -414,19 +414,10 @@ def _run_loop(self, items):
             results.append(res)
             del task_vars[loop_var]
 
-            # clear 'connection related' plugin variables for next iteration
-            if self._connection:
-                clear_plugins = {
-                    'connection': self._connection._load_name,
-                    'shell': self._connection._shell._load_name
-                }
-                if self._connection.become:
-                    clear_plugins['become'] = self._connection.become._load_name
-
-                for plugin_type, plugin_name in iteritems(clear_plugins):
-                    for var in C.config.get_plugin_vars(plugin_type, plugin_name):
-                        if var in task_vars and var not in self._job_vars:
-                            del task_vars[var]
+            # needed for winrm/pspr before clearing the connection
+            self._connection.close()
+            # clear the connection for the next iteration that may have different options set
+            self._connection = None
```

Comment:
giving it more thought, the fix here might be to just remove the old code 9it is really not doing what it is supposed to, task_vars is not updated anymore

then 'real fix' for the become issue is something like self.prompt = CLS.prompt in set_options before processing rest

Reply:
> giving it more thought, the fix here might be to just remove the old code 9it is really not doing what it is supposed to, task_vars is not updated anymore

We still need the original code to account for https://github.com/ansible/ansible/issues/58876:

```yaml
- hosts: test
  gather_facts: no
  tasks:
  - ping:
    remote_user: "{{ item }}"
    with_items:
      - centos
      - ec2-user
      - root
```

> then 'real fix' for the become issue is something like self.prompt = CLS.prompt in set_options before processing rest

So it sounds like we are back to https://github.com/ansible/ansible/pull/59764 although we need to we override `set_options` in each become plugin since the default value of `prompt` differs per plugin.